From 4def2abeeedac0ca9f494ebbc1dff0197934fcf3 Mon Sep 17 00:00:00 2001
From: Kareti Pavani Sireesha Reddy <kpavanis@hu-kpavanis-blr.qualcomm.com>
Date: Wed, 25 Feb 2026 14:22:33 +0530
Subject: [PATCH] Subject: [PATCH] Adding channels to the preference report
 only if they are not in statically non operable channels list.

Issue: Prplmesh is selecting wrong channel from channel selection report.

Root Cause: Even statically non operable channels are added as operable in
preference report.

Signed-off-by: Kareti Pavani Sireesha Reddy <kpavanis@hu-kpavanis-blr.qualcomm.com>
---
 .../slave/tasks/channel_selection_task.cpp    | 20 ++++++++++++++++---
 .../slave/tasks/channel_selection_task.h      |  1 +
 agent/src/beerocks/slave/tlvf_utils.cpp       | 10 ++++++++++
 agent/src/beerocks/slave/tlvf_utils.h         |  4 ++++
 4 files changed, 32 insertions(+), 3 deletions(-)

diff --git a/agent/src/beerocks/slave/tasks/channel_selection_task.cpp b/agent/src/beerocks/slave/tasks/channel_selection_task.cpp
index b053075e7..7a9ba853c 100644
--- a/agent/src/beerocks/slave/tasks/channel_selection_task.cpp
+++ b/agent/src/beerocks/slave/tasks/channel_selection_task.cpp
@@ -1178,6 +1178,23 @@ bool ChannelSelectionTask::build_channel_preference_report(const sMacAddr &radio
             // Operating classes 128-130,132-135 use center channel **unlike the other classes**,
             // so convert center channel and bandwidth to main channel.
             // For more info, refer to Table E-4 in the 802.11 specification.
+            AgentDB::sChannelPreference preference_key(
+                    oper_class_num, wfa_map::cPreferenceOperatingClasses::ePreference::NON_OPERABLE,
+                    wfa_map::cPreferenceOperatingClasses::eReasonCode::UNSPECIFIED);
+            // Access non-operable channels for a specific operating class
+            std::lock_guard<std::mutex> lock(tlvf_utils::operating_class_nonoper_channels_mutex);
+            auto it = tlvf_utils::operating_class_nonoper_channels.find(oper_class_num);
+            if (it != tlvf_utils::operating_class_nonoper_channels.end()) {
+                const auto& channels = it->second;
+                bool is_non_oper = channels.find(channel_of_oper_class) != channels.end();
+
+                if (is_non_oper) {
+                    LOG(INFO) << "Channel " << int(channel_of_oper_class) << " is NON-operable in operating class "
+                        << int(oper_class_num);
+                    radio->channel_preferences[preference_key].insert(channel_of_oper_class);
+                    continue;
+                }
+            }
             const auto beacon_channels =
                 son::wireless_utils::is_operating_class_using_central_channel(oper_class_num)
                     ? son::wireless_utils::center_channel_to_beacon_channels(
@@ -1186,9 +1203,6 @@ bool ChannelSelectionTask::build_channel_preference_report(const sMacAddr &radio
                     : std::vector<uint8_t>{channel_of_oper_class};
 
             // Assume non-operable
-            AgentDB::sChannelPreference preference_key(
-                oper_class_num, wfa_map::cPreferenceOperatingClasses::ePreference::NON_OPERABLE,
-                wfa_map::cPreferenceOperatingClasses::eReasonCode::UNSPECIFIED);
             for (const auto beacon_channel : beacon_channels) {
                 auto tmp_preference_key =
                     get_preference_key(beacon_channel, oper_class_num, oper_class_bw);
diff --git a/agent/src/beerocks/slave/tasks/channel_selection_task.h b/agent/src/beerocks/slave/tasks/channel_selection_task.h
index 42581e7c7..d53004146 100644
--- a/agent/src/beerocks/slave/tasks/channel_selection_task.h
+++ b/agent/src/beerocks/slave/tasks/channel_selection_task.h
@@ -20,6 +20,7 @@
 #include <tlvf/wfa_map/tlvSpatialReuseRequest.h>
 #include <tlvf/wfa_map/tlvTransmitPowerLimit.h>
 #include <unordered_map>
+#include <algorithm>
 
 #include <beerocks/tlvf/enums/eDfsState.h>
 
diff --git a/agent/src/beerocks/slave/tlvf_utils.cpp b/agent/src/beerocks/slave/tlvf_utils.cpp
index bd05d6d6b..9adb708a2 100644
--- a/agent/src/beerocks/slave/tlvf_utils.cpp
+++ b/agent/src/beerocks/slave/tlvf_utils.cpp
@@ -18,9 +18,12 @@
 #include <tlvf/wfa_map/tlvSpatialReuseReport.h>
 
 using namespace beerocks;
+std::mutex tlvf_utils::operating_class_nonoper_channels_mutex;
+std::unordered_map<uint8_t, std::unordered_set<uint8_t>> tlvf_utils::operating_class_nonoper_channels;
 
 constexpr uint8_t MINIMUM_SCAN_INTERVAL_SEC = 2;
 
+
 /**
  * @brief Get the maximum transmit power of operating class.
  *
@@ -173,6 +176,13 @@ bool tlvf_utils::add_ap_radio_basic_capabilities(ieee1905_1::CmduMessageTx &cmdu
         auto non_oper_channels =
             get_operating_class_non_oper_channels(radio->channels_list, op_class);
 
+        // Store in the static map for later use
+        std::lock_guard<std::mutex> lock(operating_class_nonoper_channels_mutex);
+        if (!non_oper_channels.empty()) {
+            operating_class_nonoper_channels[op_class] =
+                    std::unordered_set<uint8_t>(non_oper_channels.begin(), non_oper_channels.end());
+        }
+
         // OPClass won't be added if its non-operable channels size is equal to its channels size
         if (non_oper_channels.size() ==
             son::wireless_utils::operating_classes_list.find(op_class)->second.channels.size()) {
diff --git a/agent/src/beerocks/slave/tlvf_utils.h b/agent/src/beerocks/slave/tlvf_utils.h
index da2ed68a8..5e585b437 100644
--- a/agent/src/beerocks/slave/tlvf_utils.h
+++ b/agent/src/beerocks/slave/tlvf_utils.h
@@ -14,6 +14,8 @@
 #include <tlvf/CmduMessageTx.h>
 
 #include <beerocks/tlvf/beerocks_message.h>
+#include <unordered_map>
+#include <vector>
 
 namespace beerocks {
 
@@ -51,6 +53,8 @@ public:
      */
     static bool get_channel_scan_map(beerocks::AgentDB::sRadio *radio,
                                      std::map<uint8_t, std::vector<uint8_t>> &scan_map);
+    static std::unordered_map<uint8_t, std::unordered_set<uint8_t>> operating_class_nonoper_channels;
+    static std::mutex operating_class_nonoper_channels_mutex;
 };
 
 } // namespace beerocks
-- 
2.34.1
